### PR TITLE
Add option to dump offloaded tracks as an event file

### PR DESCRIPTION
**Pull request by sethrj**
This adds a shared "offload writer" that writes one "event" per `flush` call, with all threads sharing a single output file (so that the separate events can be parallelized when reloading the file).

### CODE REVIEW

**Comment by sethrj** [comment]
@whokion Can you check out this branch, set `options.offload_output_file = "cms-offload.hepmc3";` (and you can disable the SDs to speed it up a bit) and see if you can generate offload files from 100 ttbar events with CMSSW for the HLLHC and Run 3 configurations?

**Comment by whokion** [comment]
No problem.  Just for my curiosity, is the cms-offload.hepmc3 different from `celeritas-output.json `with 
   `options_->output_file = "celeritas-output.json"`
or  root output by cmsRun, for example,  `cms-hllhc-step2.root `with
  `fileName = cms.untracked.string('file:cms-hllhc-step2.root')`
inside cms.OutputModule("PoolOutputModule", ...)?

**Comment by sethrj** [comment]
- The JSON output file is for Celeritas internals (kernel diagnostics, etc.)
- The ROOT output is for processed Geant4 physics data
- The "offload" HepMC3 output file is for the tracks that Geant4 sends to celeritas

**Comment by whokion** [comment]
cmsRun with this branch runs successfully for both Run3 and HL-LHC configurations.  Please see the size of file output with 100 ttbar events:
-rw-r--r-- 1 syjun g4gpu 3262865394 Jul 13 12:10 cms-offload-hllhc.hepmc3
-rw-r--r-- 1 syjun g4gpu 1136394364 Jul 13 11:22 cms-offload-run3.hepmc3

**Comment by whokion** [comment]
Temporarily, can down those output files by
wget https://g4cpt.fnal.gov/g4p/download/celeritas/cms-offload-run3.hepmc3 
wget https://g4cpt.fnal.gov/g4p/download/celeritas/cms-offload-hllhc.hepmc3

**Comment by sethrj** [comment]
@whokion Thanks for those files; I've downloaded them so you can delete if you want to clean up space.

File: src/accel/SharedParams.hh [hunk 1]
```diff
@@ -84,19 +85,26 @@ class SharedParams
...
   private:
     //// DATA ////
 
     std::shared_ptr<CoreParams> params_;
-    std::shared_ptr<detail::HitManager> hit_manager_;
+    SPHitManager hit_manager_;
     std::shared_ptr<StepCollector> step_collector_;
     VecG4ParticleDef particles_;
     std::string output_filename_;
+    std::shared_ptr<detail::OffloadWriter> offload_writer_;
```
Comment (amandalund) [inline src/accel/SharedParams.hh]:
```suggestion
    SPOffloadWriter offload_writer_;
```

**Review by amandalund — APPROVED** [review]
Thanks @sethrj!

One minor observation: right now all the events in the output have event ID 0. This won't matter for us since we set our own sequential event IDs in the HepMC3 reader, but we might still want to add a `evt.set_event_number(event_count_)` in the `EventWriter`.

File: src/accel/SharedParams.hh (hunk 1 above)
Comment (sethrj) [reply src/accel/SharedParams.hh]: Good point, we definitely want to do that! I didn't notice that property.

**Comment by sethrj** [comment]
I also realized that `HepMC3PrimaryGenerator` is giving all vertices the "event" position so I'll update and test that as well.

**Comment by sethrj** [comment]
I'm going to actually move the hepmc3/io fix to a separate PR since that's a problem that exists on the v0.3 release.

**Pull request merged**